src/irmin-pack/inode.ml

```diff
@@ -722,6 +735,115 @@ struct
           pre_hash_node (Node.v vs)
       in
       Irmin.Type.map I.t ~pre_hash (fun v -> { find = niet; v }) (fun t -> t.v)
+
+    module Private = struct
+      let hash t = I.hash t.v
+      let stable t = I.stable t.v
+      let length t = I.length t.v
+
+      let ht = Hashtbl.create 19
+
+      let pp_semi ppf _ = Format.pp_print_string ppf ";"; Format.pp_print_cut ppf ()
+
+      let pp_hash_hr =
+        let cpt = ref 0 in
+        let new_id () =
+          incr cpt;
+          Fmt.str "hash_%d" !cpt
+        in
+        fun ppf h ->
+          let id = match Hashtbl.find_opt ht h with
+            | None ->
+              let id = new_id () in
+              Hashtbl.add ht h id;
+              id
+            | Some id -> id
+          in Fmt.pf ppf "%s" id
+
+      let pp_field l prj pp_v ppf v =
+        Fmt.(pf ppf "@[<h>%s =@ %a@]" l pp_v (prj v))
+
+      let pp_array ?(pp_sep=pp_semi) pp_elt ppf a =
+          Fmt.pf ppf "@[<v 0>@[<v -8>[@,";
+          for i = 0 to Array.length a - 2 do
+            Fmt.pf ppf "%a%a" pp_elt a.(i) pp_sep ()
+          done;
+          if Array.length a > 0 then
+            Fmt.pf ppf "%a" pp_elt a.(Array.length a - 1);
+          Fmt.pf ppf "@]@]@,]"
+
+      let pp_record ?(pp_sep=pp_semi) l ppf r =
```

Comment:
You can do something similar with a combination of `Fmt.Dump.record` and `Fmt.Dump.field`

Reply:
The problem I encountered with `Fmt.Dump.array/record` is that it's forcing a normal box surrounding a vertical box with an indentation of 0 and I couldn't manage to get some control over it. The first version I wrote was only using `Fmt.Dump` but I was not satisfied with the result (this may be a result of me just discovering Fmt as I used Format for a long time now and not using it properly). The `field` function forces a normal box with an indentation of 1 and prints all the fields between quotation marks.

As an example, when I pretty-printed an inode with `Fmt.Dump` I got the following result:

```
{
  "hash":"9e455b571805ec9d3404d9ac86009757382db687",
  "stable":true,
  "v":{"Inodes":{"seed":0,
       "length":3,
       "entries":[...
```
While I'd rather have either

```
{
  hash = 9e455b571805ec9d3404d9ac86009757382db687;
  stable = true;
  Inodes = { seed = 0;
             length = 3;
             entries = [...
```
for a wide version or
```
{
  hash = 9e455b571805ec9d3404d9ac86009757382db687;
  stable = true;
  Inodes = (
    seed = 0;
    length = 3;
    entries = [...
```
for a more compact one.
It allows me to have indent guides in emacs: 

![image](https://user-images.githubusercontent.com/5543639/104766768-699dd280-576b-11eb-9a8c-633a47cb112e.png)

Do you know if I can achieve this kind of result with Fmt?

Reply:
How did you get the first result? It seems you are printing the JSON representation of the value.

You can use `Fmt.string` as a parameter of `Fmt.Dump.record` if you don't like quotes. For instance:

```ocaml
open Fmt.Dump

let pp_field k f v = Fmt.Dump.field ~label:Fmt.string k f v

let pp =
  record [
      pp_field "foo" (fun _ -> "x") Fmt.string;
      pp_field "bar" (fun _ -> 3) Fmt.int
    ]

let () = Fmt.epr "XXX %a\n%!" pp ()
```

will print:
```
XXX { foo = x;
      bar = 3 }
```

For instance, that's exactly what `Irmin.Type.pp` [is doing](https://github.com/mirage/repr/blob/main/src/repr/type_pp.ml#L95-L100) so you could also just directly use `Irmin.Type.pp t`

Reply:
I got the first result by applying naively all the `Fmt.Dump.functions` that could fit my needs to an inode :D
Thanks for the `~label:Fmt.string` trick ;-)

There's still this first field on the `{` line that bothers me though but if it just bothers me I'll happily go back to Fmt and use `Irmin.Type.pp` (actually, thinking about it, it would make the pretty-printing homogeneous with the rest of Irmin so I should use it. I'll change it monday!) 

Thanks for the review! :-)

Reply:
(Feel free to submit a PR that improves the boxing of [`repr` pretty printers](https://github.com/mirage/repr/blob/main/src/repr/type_pp.ml) too, if you find they're not quite ideal. That way we can benefit from this everywhere. I think the current choices were fairly ad-hoc, so a pass would be very appreciated.)

Reply:
Ok, with great pleasure :-)